### PR TITLE
Mzk 41 layout7 cpu

### PR DESCRIPTION
**Pull request by zk-steve**
# Check List

- [x] Don't forget to squash commits into meaningful chunks before merging
- [x] Check every test passed.
- [x] Did you split imports into std and custom parts?
- [x] Have you updated the consts in ORN?
- [x] Have you run ORN with the latest version?
- [x] Check your commit messages.
- [x] Have you added meaningful comments?
- [x] Don't forget to optimize gas cost!
  - [x] Calling vector::length() multiple times is expensive (e.g. in loops), save it to a variable instead.
  - [x] Consider using consts for things that can be pre-computed.
  - [x] `pow(2, x) => (1 << x+1)`
  - [x] `inline` short functions to save gas.
  - [x] Find all vector::empty() in all files and replace them with the first append
  - [x] Use vector::\*\_reverse\_\*() version to save gas.
- [x] Check Visibility of functions

- - -
This change is [Reviewable](https://reviewable.io/reviews/sota-zk-labs/navori/14)

### CODE REVIEW

**Comment by zk-steve** [comment]
rebase

File: sources/cpu/layout7/cpu_constraint_poly.move [hunk 1]
```diff
@@ -0,0 +1,2281 @@
...
+
+        for (i in 0..remain) {
+            push_back(&mut ctx, 0);
+        };
+
+        {
+            // compute expmods
+            // expmods[0] = point^(trace_length / 2048)
+            {
+                let val = fexp(/*point*/ *borrow(&ctx, 34), (/*trace_length*/ *borrow(&ctx, 7) / 2048));
```
Comment (zk-steve) [inline sources/cpu/layout7/cpu_constraint_poly.move:32]: @0x5ea000000 can we create variables for *borrow(&ctx, 34), *borrow(&ctx, 7)? These values are used multiple times

File: sources/cpu/layout7/cpu_constraint_poly.move [hunk 2]
```diff
@@ -0,0 +1,2281 @@
...
+
+    public fun fexp(a: u256, b: u256): u256 {
+        0
+    }
+
+    public fun fallback(ctx: &vector<u256>): u256 {
+        let ctx = *ctx;
+        let res = 0;
+
+        let remain = 404 - vector::length(&ctx);
```
Comment (zk-steve) [inline sources/cpu/layout7/cpu_constraint_poly.move:22]: should init to 286 instead, and the code below use vector::push_back()

File: cpu/sources/layout7/cpu_constraint_poly.move [hunk 3]
```diff
@@ -0,0 +1,3926 @@
...
+    // [0x23a0, 0x23c0) - poseidon__poseidon__partial_rounds_state1_cubed_21
+    // [0x23c0, 0x29c0) - expmods
+    // [0x29c0, 0x2d40) - domains
+    // [0x2d40, 0x2f80) - denominator_invs
+    // [0x2f80, 0x31c0) - denominators
+    // [0x31c0, 0x3280) - expmod_context
+
+
+    public fun fallback(ctx: &vector<u256>): u256 {
+        let ctx = *ctx;
```
Comment (hduoc2003) [inline cpu/sources/layout7/cpu_constraint_poly.move:114]:
```suggestion
```

File: cpu/sources/layout7/cpu_constraint_poly.move [hunk 4]
```diff
@@ -0,0 +1,3926 @@
...
+                *borrow_mut(&mut ctx, 295) = val;
+            };
+            // expmods[10] = trace_generator^(trace_length / 32)
+            {
+                let val = fpow(/*trace_generator*/ trace_generator, (/*trace_length*/ trace_length / 32));
+                *borrow_mut(&mut ctx, 296) = val;
+            };
+            // expmods[11] = trace_generator^(3 * trace_length / 64)
+            {
+                let val = fpow(/*trace_generator*/ trace_generator, (fmul(3, /*trace_length*/ trace_length) / 64));
```
Comment (hduoc2003) [inline cpu/sources/layout7/cpu_constraint_poly.move:220]: Pre-compute `tg64 = trace_generator^(trace_length / 64)`, then `expmods[11] = fpow(tg64, 3)`. So we can reuse `tg64`.

File: cpu/sources/layout7/cpu_constraint_poly.move [hunk 5]
```diff
@@ -0,0 +1,3926 @@
...
+                *borrow_mut(&mut ctx, 297) = val;
+            };
+            // expmods[12] = trace_generator^(trace_length / 16)
+            {
+                let val = fpow(/*trace_generator*/ trace_generator, (/*trace_length*/ trace_length / 16));
+                *borrow_mut(&mut ctx, 298) = val;
+            };
+            // expmods[13] = trace_generator^(5 * trace_length / 64)
+            {
+                let val = fpow(/*trace_generator*/ trace_generator, (fmul(5, /*trace_length*/ trace_length) / 64));
```
Comment (hduoc2003) [inline cpu/sources/layout7/cpu_constraint_poly.move:230]: like [this one](https://github.com/sota-zk-labs/navori/pull/14/files#r1738112091)

File: cpu/sources/layout7/cpu_constraint_poly.move [hunk 6]
```diff
@@ -0,0 +1,3926 @@
...
+        let oods_point = *borrow(&ctx, 34);
+
+        let point = oods_point;
+
+        {
+            // compute expmods
+            // expmods[0] = point^(trace_length / 2048)
+            {
+                let val = fpow(/*point*/ point, (/*trace_length*/ trace_length / 2048));
+                *borrow_mut(&mut ctx, 286) = val;
```
Comment (hduoc2003) [inline cpu/sources/layout7/cpu_constraint_poly.move:166]: use variable `ctx_286` instead of `ctx[286]`

**Review by zk-steve — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/sota-zk-labs/navori/14#-:-O6_UOV70ZRs65TdiVXz:bnfp4nl)* status: 0 of 45 files reviewed, 6 unresolved discussions (waiting on @hduoc2003)

**Review by hduoc2003 — APPROVED** [review]

File: cpu-constraint-poly/sources/cpu_constraint_poly.move [hunk 7]
```diff
@@ -0,0 +1,3950 @@
...
+    // [0x29c0, 0x2d40) - domains
+    // [0x2d40, 0x2f80) - denominator_invs
+    // [0x2f80, 0x31c0) - denominators
+    // [0x31c0, 0x3280) - expmod_context
+
+
+    public fun fallback(ctx: vector<u256>): u256 {
+        let res = 0;
+
+        let remain = 404 - vector::length(&ctx);
```
Comment (zk-steve) [inline cpu-constraint-poly/sources/cpu_constraint_poly.move:115]: 4O4 ?

**Pull request merged**